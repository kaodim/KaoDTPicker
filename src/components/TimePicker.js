import React, { Component } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import HeaderClose from './HeaderClose'
import SurchargeBanner from './SurchargeBanner'
import HourList from './HourList'

class TimePicker extends Component {
  constructor(props) {
    super(props)
    this.state = {
      bannerText: this.props.bannerText,
      selectedTime: '',
      showPrice: false,
      totalSurchargePrice: ''
    }
  }

  handleHourChange = (hour, locTotalPrice, showPrice) => {
    this.setState({ selectedTime: hour, showPrice, totalSurchargePrice: locTotalPrice })
    if (showPrice) {
      this.setState({
        bannerText: `${moment(hour).format('hh:mmA')} has a surcharge of ${locTotalPrice}`
      })
    } else {
      this.setState({
        bannerText: `${moment(hour).format('DD MMM (ddd)')} has a surcharge of ${locTotalPrice}`
      })
    }
  }

  handleDoneClick = () => {
    const timeObj = {
      date: this.state.selectedTime,
      locTotalPrice: this.state.totalSurchargePrice
    }
    this.props.onChange(timeObj)
  }

  render() {
    const { bannerText, selectedTime, showPrice, totalSurchargePrice } = this.state
    const {
      backTP,
      bannerPrice,
      closeTP,
      metaTimeSurchargable,
      selectedDate,
      surchargeGif,
      timeslots
    } = this.props
    return (
      <section className="kld-daypicker__modal">
        <div className="kld-timepicker">
          <div className="kld-timepicker__header">
            <span className="kld-timepicker__header-title">{`Select time â€¢ ${moment(
              selectedDate
            ).format('DD MMM')}`}</span>
            <HeaderClose className="kld-timepicker__header-close" onClick={closeTP} />
          </div>
          {metaTimeSurchargable && (
            <div className="kld-timepicker__banner">
              <SurchargeBanner label={bannerText} surchargeGif={surchargeGif} />
            </div>
          )}
          {/* ========== Timeslots Selection ========== */}
          <div>
            <HourList
              onChange={this.handleHourChange}
              selectedTime={selectedTime}
              timeslots={timeslots}
            />
          </div>
          <div className="kld-timepicker__footer">
            <button className="kld-timepicker__footer-btn-back" onClick={backTP}>
              Back
            </button>
            <button
              className="kld-timepicker__footer-btn-done"
              disabled={!selectedTime}
              onClick={this.handleDoneClick}>
              Done
            </button>
          </div>
        </div>
      </section>
    )
  }
}

TimePicker.defaultProps = {
  backTP: () => {},
  bannerPrice: '',
  bannerText: '',
  closeTP: () => {},
  metaTimeSurchargable: false,
  onChange: () => {},
  selectedDate: '',
  surchargeGif: '',
  timeslots: []
}

TimePicker.propTypes = {
  backTP: PropTypes.func,
  bannerPrice: PropTypes.string,
  bannerText: PropTypes.string,
  closeTP: PropTypes.func,
  metaTimeSurchargable: PropTypes.bool,
  onChange: PropTypes.func,
  selectedDate: PropTypes.string,
  surchargeGif: PropTypes.string,
  timeslots: PropTypes.array
}

export default TimePicker
