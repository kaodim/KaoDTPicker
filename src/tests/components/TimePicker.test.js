import React from 'react'
import { shallow } from 'enzyme'
import calTimeslots from '../fixtures/calTimeslots'
import TimePicker from '../../components/TimePicker'

let wrapper
const date1 = '2020-01-17T07:00:00.000+08:00'
const date2 = '2020-01-17T08:00:00.000+08:00'
const onChangeSpy = jest.fn()

describe('should render TimePicker correctly', () => {
  beforeEach(() => {
    wrapper = shallow(
      <TimePicker selectedDate={date1} timeslots={calTimeslots} onChange={onChangeSpy} />
    )
  })

  test('should match TimePicker snapshot with selectedDate', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test.skip('should show bannerPrice when there is bannerText', () => {
    wrapper.setProps({
      bannerPrice: 'RM10'
    })
    expect(wrapper.find('SurchargeBanner').length).toBe(1)
  })

  test('should call handleHourChange', () => {
    wrapper.find('HourList').prop('onChange')(date1)
    expect(wrapper.state('selectedTime')).toEqual(date1)
  })

  test('should disable Done button when timeslot has not selected yet', () => {
    expect(wrapper.find('button.kld-timepicker__footer-btn-done').props().disabled).toBe(true)
  })

  test('should call handleDoneClick when click on Done button', () => {
    const price = 'RM10'
    const onClickObj = {
      date: date2,
      locTotalPrice: price
    }
    wrapper.setState({ selectedTime: date2 })
    wrapper.setProps({ bannerPrice: price })
    wrapper.find('button.kld-timepicker__footer-btn-done').simulate('click')
    expect(onChangeSpy).toHaveBeenCalledWith(onClickObj)
  })

  test('should have defaultProps func', () => {
    wrapper.find('button.kld-timepicker__footer-btn-back').simulate('click')
    wrapper
      .find('HeaderClose')
      .props()
      .onClick('click')
    wrapper.find('button.kld-timepicker__footer-btn-back').simulate('click')
    expect(TimePicker.defaultProps.backTP).toBeDefined()
    expect(TimePicker.defaultProps.closeTP).toBeDefined()
  })
})
