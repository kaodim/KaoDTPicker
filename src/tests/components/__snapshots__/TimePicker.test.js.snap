// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TimePicker correctly should match TimePicker snapshot with selectedDate 1`] = `
<section
  className="kld-daypicker__modal"
>
  <div
    className="kld-timepicker"
  >
    <div
      className="kld-timepicker__header"
    >
      <span
        className="kld-timepicker__header-title"
      >
        Select time â€¢ 17 Jan
      </span>
      <HeaderClose
        className="kld-timepicker__header-close"
        height="24"
        onClick={[Function]}
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
    <div>
      <HourList
        onChange={[Function]}
        selectedTime=""
        timeslots={
          Array [
            Object {
              "available": true,
              "human_readable_price": "",
              "localized_total_price": "RM50",
              "price": 0,
              "show_price": false,
              "surchargable": false,
              "total_price": 50,
              "value": "2020-01-17T07:00:00.000+08:00",
            },
            Object {
              "available": true,
              "human_readable_price": "",
              "localized_total_price": "RM50",
              "price": 0,
              "show_price": false,
              "surchargable": false,
              "total_price": 50,
              "value": "2020-01-17T12:30:00.000+08:00",
            },
            Object {
              "available": true,
              "human_readable_price": "",
              "localized_total_price": "RM60",
              "price": 10,
              "show_price": true,
              "surchargable": true,
              "total_price": 60,
              "value": "2020-01-17T17:30:00.000+08:00",
            },
          ]
        }
      />
    </div>
    <div
      className="kld-timepicker__footer"
    >
      <button
        className="kld-timepicker__footer-btn-back"
        onClick={[Function]}
      >
        Back
      </button>
      <button
        className="kld-timepicker__footer-btn-done"
        disabled={true}
        onClick={[Function]}
      >
        Done
      </button>
    </div>
  </div>
</section>
`;
